core/modules/nat.h

```diff
@@ -100,38 +100,45 @@ class AvailablePorts {
  public:
   // Tracks available ports within the given IP prefix.
   explicit AvailablePorts(const CIDRNetwork &prefix)
-      : prefix_(prefix), free_list_(), next_expiry_() {
-    uint32_t min = ntohl(prefix_.addr & prefix_.mask);
-    uint32_t max = ntohl(prefix_.addr | (~prefix_.mask));
-
-    for (uint32_t ip = min; ip <= max; ip++) {
+      : prefix_(prefix),
+        buffer_(),
+        free_list_(),
+        next_expiry_(),
+        min_(),
+        max_() {
+    min_ = ntohl(prefix_.addr & prefix_.mask);
+    max_ = ntohl(prefix_.addr | (~prefix_.mask));
+
+    for (uint32_t ip = min_; ip <= max_; ip++) {
       for (uint32_t port = MIN_PORT; port <= MAX_PORT; port++) {
-        free_list_.emplace_back(htonl(ip), htons((uint16_t)port),
-                                new FlowRecord());
+        buffer_.emplace_back(FlowRecord());
       }
     }
-    std::random_shuffle(free_list_.begin(), free_list_.end());
-  }
 
-  ~AvailablePorts() {
-    for (auto &i : free_list_) {
-      FlowRecord *r = std::get<2>(i);
-      delete r;
-    }
+    free_list_.resize(buffer_.size());
+    std::iota(free_list_.begin(), free_list_.end(), 0);
+    std::random_shuffle(free_list_.begin(), free_list_.end());
   }
 
   // Returns a random free IP/port pair within the network and removes it from
   // the free list.
   std::tuple<IPAddress, uint16_t, FlowRecord *> RandomFreeIPAndPort() {
-    std::tuple<IPAddress, uint16_t, FlowRecord *> r = free_list_.back();
+    size_t i = free_list_.back();
     free_list_.pop_back();
-    return r;
+
+    uint32_t ip = i / (MAX_PORT - MIN_PORT + 1) + min_;
```

Comment:
64bit integer division/modulo operations are one of the most expensive instructions in x86. For example, it takes more than 50% of CPU time of the `Measure` module. Would it be possible to avoid using division/modulo operations somehow?